opt: remove redundant arbiter indexes

mgartner opened the pull request:
We select all unique indexes and unique constraints as arbiters for
`INSERT .. ON CONFLICT` statements without explicit conflict columns or
constraints. In some cases, unique constraint arbiters make unique index
arbiters redundant, which results in query plans that perform
unnecessary work to check for conflicts. Most notably, redundant unique
indexes are chosen as arbiters for both partitioned and hash-sharded
indexes which synthesize `UNIQUE WITHOUT INDEX` constraints.

For example, consider the table and statement which mimics a
simplified hash-sharded index:

    CREATE TABLE t (
        a INT,
        shard_a INT,
        UNIQUE INDEX shard_a_a_key (shard_a, a),
        UNIQUE WITHOUT INDEX a_key (a)
    )

    INSERT INTO t VALUES (1, 2) ON CONFLICT DO NOTHING

There is no need to use both `shard_a_a_key` and `a_key` as arbiters for
the `INSERT` statement because any conflict in `shard_a_a_key` will also
be a conflict in `a_key`. Only `a_key` is required to be an arbiter.

This commit removes these redundant arbiter indexes, which improves
query plans.

Fixes #75498

Release note (performance improvement): The optimizer produces more
efficient query plans for `INSERT .. ON CONFLICT` statements that do not
have explicit conflict columns or constraints and are performed on
partitioned tables.

mgartner committed:
opt: remove redundant arbiter indexes

We select all unique indexes and unique constraints as arbiters for
`INSERT .. ON CONFLICT` statements without explicit conflict columns or
constraints. In some cases, unique constraint arbiters make unique index
arbiters redundant, which results in query plans that perform
unnecessary work to check for conflicts. Most notably, redundant unique
indexes are chosen as arbiters for both partitioned and hash-sharded
indexes which synthesize `UNIQUE WITHOUT INDEX` constraints.

For example, consider the table and statement which mimics a
simplified hash-sharded index:

    CREATE TABLE t (
        a INT,
        shard_a INT,
        UNIQUE INDEX shard_a_a_key (shard_a, a),
        UNIQUE WITHOUT INDEX a_key (a)
    )

    INSERT INTO t VALUES (1, 2) ON CONFLICT DO NOTHING

There is no need to use both `shard_a_a_key` and `a_key` as arbiters for
the `INSERT` statement because any conflict in `shard_a_a_key` will also
be a conflict in `a_key`. Only `a_key` is required to be an arbiter.

This commit removes these redundant arbiter indexes, which improves
query plans.

Fixes #75498

Release note (performance improvement): The optimizer produces more
efficient query plans for `INSERT .. ON CONFLICT` statements that do not
have explicit conflict columns or constraints and are performed on
partitioned tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76961)
<!-- Reviewable:end -->

chengxiong-ruan reviewed (approved):
I'm very pleased reading through the pr. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-:-MwdaCtt2SYoShYqLBv1:b-o06j7c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @msirek, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdY_ZW72kdTr33krZ8:-MwdY_ZW72kdTr33krZ9:bxj9zci) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L179)):*
> ```Go
> // INDEX constraints that are synthesized for partitioned and hash-sharded
> // indexes, which the user has no control over.
> type minArbiterSet struct {
> ```

just a nit feel free to ignore, but to be honest, I'm not sure how much `singleIndexConflictOrdsCache` gonna save us from memory footprint. But if there is a strong reason that's helpful, I think it worth mentioning in comments at struct def level that either `singleIndexConflictOrdsCache` or `indexConflictOrdsCache` is set, not both. This is also relative to how `addIndex` and `addUniqueConstraint` cooperate with each other, so worth mentioning that we always finish adding index with `addIndex` always before logic adding unique constraint starts, that's why we always know if there is one single index or more when we `initCache` which is required by `addUniqueConstraint` ...... Sorry it took me quite a few minutes to fully understand, but I think the code is a good read in general.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 3630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdaWqo2jWDoINC4p42:-MwdaWqo2jWDoINC4p43:b-hklegq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/testdata/upsert#L3630)):*
> ```
> # columns or constraints.
> exec-ddl
> CREATE TABLE min_arbiters (
> ```

nit: I think it worth to break up this table into smaller test, but feel free to ignore as well because I know there'd be annoying amount of combinations >_>


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdY_ZW72kdTr33krZ8:-MwgIJiZ9RQfCp251vkB:bafas2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L179)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

just a nit feel free to ignore, but to be honest, I'm not sure how much `singleIndexConflictOrdsCache` gonna save us from memory footprint. But if there is a strong reason that's helpful, I think it worth mentioning in comments at struct def level that either `singleIndexConflictOrdsCache` or `indexConflictOrdsCache` is set, not both. This is also relative to how `addIndex` and `addUniqueConstraint` cooperate with each other, so worth mentioning that we always finish adding index with `addIndex` always before logic adding unique constraint starts, that's why we always know if there is one single index or more when we `initCache` which is required by `addUniqueConstraint` ...... Sorry it took me quite a few minutes to fully understand, but I think the code is a good read in general.
</blockquote></details>

Yes, after sleeping on it I agree this might be over-engineered. Even without the `singleIndexConflictOrdsCache`, we avoid allocations when the table has no `UNIQUE WITHOUT INDEX` constraints, so that's good. I'll simplify this unless someone disagrees.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-:-MwgFnid10NspoBA4uLm:bdtx8r6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgY3W449UDJ0kDkf8-:-MwgY3W449UDJ0kDkf80:bgsglhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L221)):*
> ```Go
> 		m.as.removeIndex(idx)
> 	}
> }
> ```

I couldn't find any issues from looking at the code so far, but taking a testing approach, I found a test case I'm wondering should pass:

```
CREATE TABLE uniq_constraint_and_partial_index (
  k INT PRIMARY KEY,
  a INT,
  b INT,
  UNIQUE INDEX (a) WHERE b > 11,
  UNIQUE WITHOUT INDEX (a) WHERE b < -1,
  UNIQUE WITHOUT INDEX (a) WHERE b > 0
);

INSERT INTO uniq_constraint_and_partial_index VALUES (2, 2, 2)
ON CONFLICT (a) WHERE b > 10 DO NOTHING;

error (42P10): there is no unique or exclusion constraint matching the ON CONFLICT specificationerror (42P10): there is no unique or exclusion constraint matching the ON CONFLICT specification
```
If the table has only the last unique index, the insert works:
```
CREATE TABLE uniq_constraint_and_partial_index (
  k INT PRIMARY KEY,
  a INT,
  b INT,
  UNIQUE WITHOUT INDEX (a) WHERE b > 0
);
```
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgFsvJ-dNshqpbm8nd:-MwgFsvJ-dNshqpbm8ne:bnikx59) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L270)):*
> ```Go
> 	// Do nothing if the cache has already been initialized.
> 	if m.singleIndexConflictOrdsCache.Empty() && m.indexConflictOrdsCache != nil {
> 		return
> ```

nit: Based on the comment, should the condition be `if !m.singleIndexConflictOrdsCache.Empty() || m.indexConflictOrdsCache != nil {`  ?
___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgLjg6EaHlsQZ1MXqd:-MwgLjg6EaHlsQZ1MXqe:b-hdi93n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L237)):*
> ```Go
> 		// the unique constraint ordinals and determining set equality.
> 		if uniqueConstraint.ColumnCount() != conflictOrds.Len() {
> 			continue
> ```

nit: It's not part of this PR, but this `if` is a duplicate of the lines of code above it.


<!-- Sent from Reviewable.io -->

msirek reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-:-MwgmAUxE1YoI2y6BM00:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgY3W449UDJ0kDkf8-:-MwgllRpAbdv-IhpU_Ah:b3dzhh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L221)):*
> CREATE TABLE uniq_constraint_and_partial_index (
>   k INT PRIMARY KEY,
>   a INT,
>   b INT,
>   UNIQUE INDEX (a) WHERE b > 11,
>   UNIQUE WITHOUT INDEX (a) WHERE b < -1,
>   UNIQUE WITHOUT INDEX (a) WHERE b > 0
> );

I found the test case fails as an optbuilder test run in Goland, but succeeds when run against a demo cluster, both on the `master` and `remove-unnecessary-arbiters` branches.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Awesome!

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-:-MwgqHMs19RaGMGsqg03:bswp6el)* status: :shipit: complete! 3 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, and @RaduBerinde)

___
*[-- commits, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgmpg_-zLNSv66SX4n):*
nit: column b came out of nowhere (here and in the PR message)
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdY_ZW72kdTr33krZ8:-MwgnWFa7l9JcO_LGiQl:bjjzafp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L179)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Yes, after sleeping on it I agree this might be over-engineered. Even without the `singleIndexConflictOrdsCache`, we avoid allocations when the table has no `UNIQUE WITHOUT INDEX` constraints, so that's good. I'll simplify this unless someone disagrees.
</blockquote></details>

Go for it
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgo4O_0fLQNUJ3O9GH:-Mwgo4O_0fLQNUJ3O9GI:bnydrsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L229)):*
> ```Go
> 
> // findRedundantIndex returns the first arbiter index which is made redundant by
> // the unique constraint. An arbiter index is redundant if:
> ```

nit: it's not clear if this list is combined with AND or OR. You could say "... is redundant if both of the following hold:"
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgocRbDlyG6SgdO9BE:-MwgocRbDlyG6SgdO9BF:bcs37bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L280)):*
> ```Go
> 		return
> 	}
> 	// Otherwise, cache each indexes conflict columns in a map.
> ```

nit: indexes -> index's
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgp-TC8ZAkmB73fMKC:-Mwgp-TD1r-e8ta7-NK1:bdtzdu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L281)):*
> ```Go
> 	}
> 	// Otherwise, cache each indexes conflict columns in a map.
> 	m.indexConflictOrdsCache = make(map[cat.IndexOrdinal]util.FastIntSet)
> ```

nit: you can do `make(map[cat.IndexOrdinal]util.FastIntSet, m.as.indexes.Len())`


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: always show arbiters in opt trees

Arbiter indexes and unique constraints are now shown in optimizer trees,
regardless of format flags.

Release note: None

mgartner reviewed (commented):
Thanks for all the great feedback!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-:-Mwh52gv2VTQAP81dj06:b5edqia)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @chengxiong-ruan, @msirek, @RaduBerinde, and @rytaft)

___
*[-- commits, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgmpg_-zLNSv66SX4n):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: column b came out of nowhere (here and in the PR message)
</blockquote></details>

Fixed.
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdY_ZW72kdTr33krZ8:-Mwh225aCgMEr4QXTMan:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L179)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Go for it
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgY3W449UDJ0kDkf8-:-Mwh0exc0DZfg-GAUTqk:bxlge4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L221)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

> CREATE TABLE uniq_constraint_and_partial_index (
>   k INT PRIMARY KEY,
>   a INT,
>   b INT,
>   UNIQUE INDEX (a) WHERE b > 11,
>   UNIQUE WITHOUT INDEX (a) WHERE b < -1,
>   UNIQUE WITHOUT INDEX (a) WHERE b > 0
> );

I found the test case fails as an optbuilder test run in Goland, but succeeds when run against a demo cluster, both on the `master` and `remove-unnecessary-arbiters` branches.
</blockquote></details>

Thanks for finding this! I've created an issue to track it: https://github.com/cockroachdb/cockroach/issues/76994
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgo4O_0fLQNUJ3O9GH:-Mwh13eXEUXkc-SnRvoa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L229)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: it's not clear if this list is combined with AND or OR. You could say "... is redundant if both of the following hold:"
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgFsvJ-dNshqpbm8nd:-Mwh18alBSOkbAl1Nk9c:bdct4qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L270)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

nit: Based on the comment, should the condition be `if !m.singleIndexConflictOrdsCache.Empty() || m.indexConflictOrdsCache != nil {`  ?
</blockquote></details>

Good catch. I've removed `singleIndexConflictOrdsCache` entirely.
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgocRbDlyG6SgdO9BE:-Mwh1G3k50FH1WmeN3yl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L280)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: indexes -> index's
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/arbiter_set.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-Mwgp-TC8ZAkmB73fMKC:-Mwh1KZF-U_yKHWI7Ced:boj1qgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/arbiter_set.go#L281)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: you can do `make(map[cat.IndexOrdinal]util.FastIntSet, m.as.indexes.Len())`
</blockquote></details>

Good catch. Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder_arbiter.go, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwgLjg6EaHlsQZ1MXqd:-Mwh1jqABGQYPBksL_SX:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/mutation_builder_arbiter.go#L237)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

nit: It's not part of this PR, but this `if` is a duplicate of the lines of code above it.
</blockquote></details>

Good catch, fixed.
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 3630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdaWqo2jWDoINC4p42:-MwgXG8U3UZeM6cPL3qv:b3b6q3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/testdata/upsert#L3630)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

nit: I think it worth to break up this table into smaller test, but feel free to ignore as well because I know there'd be annoying amount of combinations >_>
</blockquote></details>

Good idea. I've split this up into multiple tests. I've also added a new commit that allows the arbiters to be shown with less-verbose opt trees, since we only really care about `arbiter indexes` and `arbiter constraints` in these tests.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @chengxiong-ruan, @msirek, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 3630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76961#-MwdaWqo2jWDoINC4p42:-Mwh94qGDzz2FzZPU2H9:b-7m74ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e5494f5d2c68983a0d1a3eb0c1a44f1182b3f5/pkg/sql/opt/optbuilder/testdata/upsert#L3630)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Good idea. I've split this up into multiple tests. I've also added a new commit that allows the arbiters to be shown with less-verbose opt trees, since we only really care about `arbiter indexes` and `arbiter constraints` in these tests.
</blockquote></details>

nice!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 26 of 26 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76961)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @chengxiong-ruan, @msirek, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

mgartner commented on pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_query_behavior:
```diff
@@ -329,6 +329,56 @@ ORDER BY pk LIMIT 5] OFFSET 2
+  VALUES ('ca-central-1', 7, 7, 8, 9) ON CONFLICT DO NOTHING
+] OFFSET 2
+----
+·
```
This plan is much better than before removing redundant arbiters. It was:

```
·
• insert
│ into: regional_by_row_table(pk, pk2, a, b, j, crdb_region)
│ arbiter indexes: regional_by_row_table_pkey, regional_by_row_table_b_key, uniq_idx
│ arbiter constraints: regional_by_row_table_pkey, regional_by_row_table_b_key, uniq_idx
│
└── • render
    │
    └── • distinct
        │ distinct on: arbiter_uniq_idx_distinct
        │ nulls are distinct
        │
        └── • render
            │
            └── • distinct
                │ distinct on: arbiter_uniq_idx_distinct
                │ nulls are distinct
                │
                └── • render
                    │
                    └── • lookup join (anti)
                        │ table: regional_by_row_table@uniq_idx (partial index)
                        │ lookup condition: (column4 = a) AND (crdb_region IN ('ap-southeast-2', 'ca-central-1', 'us-east-1'))
                        │ pred: column5 > 0
                        │
                        └── • lookup join (anti)
                            │ table: regional_by_row_table@regional_by_row_table_b_key
                            │ equality cols are key
                            │ lookup condition: (column5 = b) AND (crdb_region IN ('ap-southeast-2', 'ca-central-1', 'us-east-1'))
                            │
                            └── • lookup join (anti)
                                │ table: regional_by_row_table@regional_by_row_table_pkey
                                │ equality cols are key
                                │ lookup condition: (column2 = pk) AND (crdb_region IN ('ap-southeast-2', 'ca-central-1', 'us-east-1'))
                                │
                                └── • lookup join (anti)
                                    │ table: regional_by_row_table@uniq_idx (partial index)
                                    │ equality: (column1, column4) = (crdb_region,a)
                                    │ pred: column5 > 0
                                    │
                                    └── • lookup join (anti)
                                        │ table: regional_by_row_table@regional_by_row_table_b_key
                                        │ equality: (column1, column5) = (crdb_region,b)
                                        │ equality cols are key
                                        │
                                        └── • cross join (anti)
                                            │
                                            ├── • values
                                            │     size: 6 columns, 1 row
                                            │
                                            └── • scan
                                                  missing stats
                                                  table: regional_by_row_table@regional_by_row_table_pkey
                                                  spans: [/'ca-central-1'/7 - /'ca-central-1'/7]
```

mgartner commented:
TFTRs!

bors r+

mgartner commented:
Failed due to flake.

bors r+
